Fix CMake build of TDNF

Conan-Kudo opened the pull request:
This pull request addresses the issues I discovered while updating packaging for TDNF to the latest revision in the dev branch.

The major changes are:

* CMake is now using `GNUInstallDirs` so that the correct target directories are used on various platforms
* Header files are now being installed so that other projects can link to TDNF
* Project versioning is now set through CMake so there's a proper fallback when Git data isn't available
* The pkgconfig files have been restored and are configured and installed by CMake
* Fix includes in Python binding sources so that it builds on a system that doesn't already have TDNF installed in the build environment
* CI has been reworked to add a Fedora-based build and build the Python module

This should fix all the issues I've discovered so far...

Conan-Kudo committed:
cmake: Install headers and use GNUInstallDirs for correct install paths

The TDNF libraries were getting installed, but they were unusable because
the header files were not being installed along with the libraries.

Additionally, the installation paths were wrong for 64-bit Linux systems
because it forced the usage of /lib where /lib64 was required. Switching
to having the installation paths defined by GNUInstallDirs by default
(included in CMake since 2.8.5) fixes this problem.

Signed-off-by: Neal Gompa <ngompa13@gmail.com>

Conan-Kudo committed:
cmake: Set up project and library versioning and install pkgconfig files

Since we're using a modern version of CMake (version 3.0+), we can
set the version as part of the project properties.

As part of doing so, the new project version setting is propagated
down to the TDNF library build targets, so that the version and
soversion are based on the project version.

Additionally, we can now have a useful fallback value for the
BuildMetaData CMake function so that builds from release tarballs
will not have broken version information.

Finally, we use these changes as part of restoring the pkgconfig
files so that they are configured and installed properly with the
correct values.

Signed-off-by: Neal Gompa <ngompa13@gmail.com>

ppadmavilasom commented:
@Conan-Kudo thanks! appreciate your help

Conan-Kudo committed:
python: Fix includes so that Python bindings build properly

Compilation was failing because /usr/include/tdnf/tdnf.h does not
exist during the build and tdnfbase.h was not being included
by its corresponding source file.

Signed-off-by: Neal Gompa <ngompa13@gmail.com>

ppadmavilasom commented:
@Conan-Kudo if we add an rpm build and an isolated dependent installation based on the built rpm, we can identify the pkg-config issues. do you see other test gaps here?

Conan-Kudo commented:
@ppadmavilasom I think that should mostly cover it. If we add a build using Fedora in here, for example, I think we can make sure things don't break like this again.

ppadmavilasom commented:
> @ppadmavilasom I think that should mostly cover it. If we add a build using Fedora in here, for example, I think we can make sure things don't break like this again.

Sure. we will add a build using fedora in addition to the one that uses photon now (as well as adding a dependency test via rpm installs or  client code compile from a newly installed chroot)

Conan-Kudo committed:
ci: Refactor CI setup to add Fedora build in addition to Photon build

In order to (hopefully) prevent future breakage, the CI has been refactored
so that TDNF is now built on both Fedora and Photon. This ensures
that TDNF remains buildable and somewhat usable on Fedora and other
similar RPM-based distributions. CI now builds the Python module to
verify that builds properly, too.

As the tests only work properly on Photon OS for now, the CI is configured
to only execute them for the Photon environment.

Signed-off-by: Neal Gompa <ngompa13@gmail.com>

Conan-Kudo commented:
@ppadmavilasom I've added a Fedora CI build of TDNF, though I have not added an RPM spec yet...

ppadmavilasom commented on ci/Dockerfile.fedora:
```diff
@@ -0,0 +1,12 @@
+MAINTAINER csiddharth@vmware.com
+
+
+RUN dnf -y upgrade
```
-q as well please

Conan-Kudo replied:
Why do quiet? Wouldn't that make it look like the CI is doing nothing for a long period of time?

Conan-Kudo replied:
Anyway, I added the quiet option.

ppadmavilasom reviewed (approved):
LGTM. Thanks for your help.

Conan-Kudo committed:
Update README to reflect change to CMake and new location of Dockerfile

Signed-off-by: Neal Gompa <ngompa13@gmail.com>

Conan-Kudo commented:
@ppadmavilasom This should be good to merge now. We have significantly more coverage than we did before, which is a major step up. There's certainly other improvements to make, but let's get this merged so that it doesn't keep getting bigger. 😄
